bluescript-reference.md

```diff
@@ -1,1133 +1,1181 @@
+opacity | 1 | In the range [0, 1] where 0 is fully transparent and 1 is fully opaque
 
-Behaviors - the **step** property that contains all behaviors for each element of the step - is an object with keys that match element IDs:
-* **\<STRING ID\>** - ID of the BlueScript element that the child behaviors are associated with
-  and has an array value that defines the behaviors of the object for each event (see below).
+**Note:** More properties are available depending on the type of BlueScript element, detailed in 
+the [element reference section](#bluescript-element-reference) below.
 
-Property | Default | Description
---- | --- | ---
-Element ID | {} | An object that defines the behavior on events. Each base behavior contains an array of actions.
+#### Behaviors
 
-<details>
-<summary>For example:</summary>
+A step's "behaviors" property contains all behaviors for each element of the step - is an object with keys that match element names to a map of event handlers, where each event handler maps to an array of [behavior actions](#behavior-actions) to execute when the event occurs. BlueScript elements can define any number of behavior actions for any number of events.
 
-```
+For example:
+```json
 {
   "behaviors": {
-    "fooLabel": {
-      <BEHAVIORS>
+    "button1": {
+      "appear": [ <ACTIONS> ],
+      "disappear": [ <ACTIONS> ],
+      "onSelect": [ <ACTIONS> ]
     }
   }
 }
 ```
-</details>
 
-#### Base Behavior Properties
+#### Functions
 
-For each event, an array of **actions** can be defined. BlueScript elements can define any number of behavior actions for any number of events.
+A step's "functions" property contains all reusable functions, keyed by function name, that can be called from other event action execution, expressions, and even other functions.
 
-Property | Default | Description
---- | --- | ---
-behavior | [ ] | Defines the list of behavior actions to be triggered on specific events. List of event names to triggered behavior action lists.
+One uses the `invoke` behavior action or expression to execute functions. Arguments are passed by name, and are referred to with the function via the `arg` expression.
+
+The `return` host action can be used to return from the function immediately.
+
+Local variable values can be assigned via the `local` assign action. Such variables exist only during the current invocation's execution.
 
-<details>
 <summary>For example:</summary>
 
-```
+```json
 {
+  "elements": [...],
+  "functions": {
+    "testFunction": [
+      {"host": "assign", "local": "v", "value": {"arg": "x"}},
+      {"host": "return", "value": {"+": [{"local": "v"}, {"local": "v"}]}}
+    ]
+  },
   "behaviors": {
-    "fooLabel": {
-      "appear": [ <ACTIONS> ],
-      "disappear": [ <ACTIONS> ]
+    "testButton": {
+      "appear": [
+        {"host": "invoke", "function": "testFunction", "args": {"x": 123}},
```

Comment:
Yes, pass by reference is what is used, matching JS semantics and much easier to implement.
I would guess Roku does the same. @addisonm can you verify?

I will add a note about this in the docs. 

I believe this is what is in fact desired, so that we can indeed change subfields or array elements of a shared object or array.